### PR TITLE
Drop support for python 3.6

### PR DESCRIPTION
**Pull request by timizuoebideri1**
### Fixes: #1268, #1435

- Drop support for python 3.6
- Update to selenium 4.x

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -119,10 +119,7 @@ netaddr = "~0.8.0"
 # Library of network related Jinja Filters
 netutils = "~1.0.0"
 # Image processing library
-Pillow = [
-    {version = "~9.0.0", python = "^3.7"},
-    {version = "~8.3.2", python = "<3.7"},
-]
+Pillow = {version = "~9.0.0", python = "^3.7"}
```
Comment (bryanculver) [inline pyproject.toml:122]:
```suggestion
Pillow = "~9.0.0"
```

Don't need the Python signifier at this point.

**Review by bryanculver — CHANGES_REQUESTED** [review]
Please review these lines as well:

https://github.com/nautobot/nautobot/blob/a190c1f64534d06daaef443d24a2a31418f5447f/pyproject.toml#L18

https://github.com/nautobot/nautobot/blob/a190c1f64534d06daaef443d24a2a31418f5447f/pyproject.toml#L34

https://github.com/nautobot/nautobot/blob/a190c1f64534d06daaef443d24a2a31418f5447f/pyproject.toml#L36-L39

https://github.com/nautobot/nautobot/blob/a190c1f64534d06daaef443d24a2a31418f5447f/pyproject.toml#L190

File: pyproject.toml [hunk 2]
```diff
@@ -119,10 +119,7 @@ netaddr = "~0.8.0"
 # Library of network related Jinja Filters
```
Comment (glennmatthews) [inline pyproject.toml:119]: Please also evaluate whether it's possible and desirable to update to the latest versions of `GitPython` and `jsonschema`, as both are currently pinned to older versions due to the python 3.6 dependency.

File: pyproject.toml [hunk 3]
```diff
@@ -119,10 +119,7 @@ netaddr = "~0.8.0"
...
 netutils = "~1.0.0"
 # Image processing library
-Pillow = [
-    {version = "~9.0.0", python = "^3.7"},
-    {version = "~8.3.2", python = "<3.7"},
-]
+Pillow = {version = "~9.0.0", python = "^3.7"}
 # PostgreSQL database adapter
 # NOTE: psycopg3 is avaiable now and nominally replaces psycopg2
 psycopg2-binary = "~2.9.3"
```
Comment (glennmatthews) [inline pyproject.toml:125]: Also please check dev dependencies - we should be able to update to the latest available releases of `black` and `coverage` now; possibly also `selenium` 4.x?

**Comment by timizuoebideri1** [comment]
> Please review these lines as well:
> 
> https://github.com/nautobot/nautobot/blob/a190c1f64534d06daaef443d24a2a31418f5447f/pyproject.toml#L190
@bryanculver should py36 be changed to py37?
Im guessing `py36` means `python 3.6`

**Comment by bryanculver** [comment]
> Please review these lines as well:
> 
> https://github.com/nautobot/nautobot/blob/a190c1f64534d06daaef443d24a2a31418f5447f/pyproject.toml#L190
> 
> 
> @bryanculver should py36 be changed to py37?
> Im guessing `py36` means `python 3.6`

Trying to look for this it's clear Black could use a bit of polish on their docs because it's opaque. Closest I found was either looking at the command line args or this issue in their repo asking for clarification:

https://github.com/psf/black/issues/751


Upon reading the thread above, I would assume this should be set to:

```toml
target_version = ['py37', 'py38', 'py39', 'py310']
```

If that presents significant changes or issues it might be best to just set it to `'py37'` but it's meant to be inclusive of all supported/targeted versions (hence the array).

File: pyproject.toml [hunk 4]
```diff
@@ -15,7 +15,6 @@ classifiers = [
     "Intended Audience :: Developers",
     "Development Status :: 5 - Production/Stable",
     "Programming Language :: Python :: 3",
-    "Programming Language :: Python :: 3.6",
     "Programming Language :: Python :: 3.7",
     "Programming Language :: Python :: 3.8",
     "Programming Language :: Python :: 3.9",
```
Comment (glennmatthews) [inline pyproject.toml:20]: Mind adding 3.10 to this list?

File: pyproject.toml [hunk 5]
```diff
@@ -90,9 +86,7 @@ djangorestframework = "~3.13.1"
 # Swagger schema generation for the REST API
 drf-yasg = {extras = ["validation"], version = "~1.20.0"}
 # Git integrations for Python
-# NOTE: GitPython 3.1.23 and later drop support for Python 3.6
-# NOTE: GitPython 3.1.19 through 3.1.23 are yanked from PyPI
-GitPython = "3.1.18"
+GitPython = "3.1.27"
```
Comment (glennmatthews) [inline pyproject.toml:89]:
```suggestion
GitPython = "~3.1.27"
```

May need to `poetry add GitPython~3.1.27` to get pyproject also updated

File: pyproject.toml [hunk 6]
```diff
@@ -176,8 +164,7 @@ mkdocs = "~1.2.3"
 # Integration Tests
 requests = "~2.27.1"
 # Selenium web drivers for live integration testing
-# NOTE: selenium 4.x drops support for python 3.6
-selenium = "~3.141.0"
+selenium = "~4.1.2"
```
Comment (glennmatthews) [inline pyproject.toml:167]: Not sure if this means we should also update the Selenium container version (https://github.com/nautobot/nautobot/blob/develop/development/docker-compose.yml#L88) as well - probably so.

File: pyproject.toml [hunk 7]
```diff
@@ -154,12 +144,10 @@ sso = ["social-auth-core"]
 
 [tool.poetry.dev-dependencies]
 # Code style enforcement
-# NOTE: black 22.1.0 stable is out, but it requires click 8.0, and celery 5.1.2 requires click 7.x.
-# NOTE: adding different Blacks as conditional on Python version fails: https://github.com/python-poetry/poetry/issues/4947
-black = {version = "~21.12b0", python = ">3.6.2"}
+black = "~22.1.0"
 # Test code coverage measurement
 # NOTE: coverage 6.3 drops support for python 3.6
```
Comment (glennmatthews) [inline pyproject.toml:149]: Can remove this comment

File: pyproject.toml (hunk 6 above)
Comment (timizuoebideri1) [reply pyproject.toml:167]: on next branch: its on latest tag
>   https://github.com/nautobot/nautobot/blob/e896524ea57aa887bfd88f0267cfbc060743929b/development/docker-compose.yml#L88

Comment (bryanculver) [reply pyproject.toml:167]: > on next branch: its on latest tag
> 
> > https://github.com/nautobot/nautobot/blob/e896524ea57aa887bfd88f0267cfbc060743929b/development/docker-compose.yml#L88

Yeah but this was addressed in `develop`. `next` needs to be brought up to date with what was just merged into `main` yesterday.

Comment (timizuoebideri1) [reply pyproject.toml:167]: 👍

**Comment by glennmatthews** [comment]
Fixes #1435 I think?

**Review by glennmatthews — APPROVED** [review]
Looks good!

**Review by bryanculver — APPROVED** [review]

**Pull request merged**